pkg/sentry/fsimpl/fuse/init.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"math"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/sentry/kernel"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+)
+
+// consts used by FUSE_INIT negotation to have the expected behavior of the existing FUSE protocol.
+const (
+	// FUSE_MAX_MAX_PAGES is the maximum value for MaxPages received in InitOut.
+	FUSE_MAX_MAX_PAGES = 256
+
+	// Maximum value for the time granularity for file time stamps, 1s.
+	fuseMaxTimeGranNs = 1000000000
+
+	// Minimum value for MaxWrite.
+	fuseMinMaxWrite = 4096
+
+	// Temporary default value for max readahead, 128kb.
+	fuseDefaultMaxReadahead = 131072
+)
+
+// Adjustable maximums for Connection's cogestion control parameters.
+// Used as the upperbound of the config values.
+// TODO: add adjust support.
+var (
+	MaxUserBackgroundRequest uint16 = FUSE_DEFAULT_MAX_BACKGROUND
+	MaxUserCongestionThreshold uint16 = FUSE_DEFAULT_CONGESTION_THRESHOLD
+)
+
+// Init sends a FUSE_INIT request, waits for the reply, and processes it.
+func (fs *filesystem) Init(creds *auth.Credentials, k *kernel.Kernel, pid uint32) error {
+	req, err := fs.initBuildRequest(creds, pid)
+	if err != nil {
+		return err
+	}
+
+	res, err := fs.fuseConn.CallTaskNonBlock(req)
+	if err != nil {
+		return err
+	}
+	if err := res.Error(); err != nil {
+		return err
+	}
+
+	var out linux.FUSEInitOut
+	if err := res.UnmarshalPayload(&out); err != nil {
+		return err
+	}
+
+	return fs.initProcessReply(k, creds, &out)
+}
+
+// Builds a FUSE_INIT request.
+func (fs *filesystem) initBuildRequest(creds *auth.Credentials, pid uint32) (*Request, error) {
+	in := linux.FUSEInitIn{
+		Major: linux.FUSE_KERNEL_VERSION,
+		Minor: linux.FUSE_KERNEL_MINOR_VERSION,
+		// TODO: find appropriate way to calculate this
+		MaxReadahead: fuseDefaultMaxReadahead, 
+		Flags: linux.FUSE_ASYNC_READ |
+			linux.FUSE_POSIX_LOCKS |
+			linux.FUSE_ATOMIC_O_TRUNC |
+			linux.FUSE_EXPORT_SUPPORT |
+			linux.FUSE_BIG_WRITES |
+			linux.FUSE_DONT_MASK |
+			linux.FUSE_SPLICE_WRITE |
+			linux.FUSE_SPLICE_MOVE |
+			linux.FUSE_SPLICE_READ |
+			linux.FUSE_FLOCK_LOCKS |
+			linux.FUSE_HAS_IOCTL_DIR |
+			linux.FUSE_AUTO_INVAL_DATA |
+			linux.FUSE_DO_READDIRPLUS |
+			linux.FUSE_READDIRPLUS_AUTO |
+			linux.FUSE_ASYNC_DIO |
+			linux.FUSE_WRITEBACK_CACHE |
```

Comment:
Do we enumerate all supported flags here? I think we need to specify a minimal set which is required. Then we will implement a feature and add its flag.

Reply:
This mimic the exact behavior of current linux kernel. I'm not sure how to tell which is a minimal set?

Reply:
The linux kernel supports all these fuse features, gVisor doesn't. I think you need to look at each of these flags and try to understand what they mean. Each flag which requires to handle one or more specific fuse command has to be dropped.

Reply:
#3199